wfl/calculators/vasp.py

```diff
@@ -4,185 +4,216 @@
 """
 
 import os
-import pathlib
+from pathlib import Path
 import tempfile
-import warnings
+import shutil
+
 import json
+from copy import deepcopy
 
 import numpy as np
 
 from ase.atoms import Atoms
-from ase.calculators.vasp.vasp import Vasp
+from ase.calculators.calculator import all_changes
+import ase.calculators.vasp.vasp
 
-from .utils import clean_rundir, handle_nonperiodic, save_results
-from ..utils.misc import atoms_to_list
+from .utils import clean_rundir, handle_nonperiodic
 
 # NOMAD compatible, see https://nomad-lab.eu/prod/rae/gui/uploads
-__default_keep_files = ["POSCAR", "INCAR", "KPOINTS", "OUTCAR", "vasprun.xml", "vasp.out"]
-__default_properties = ["energy", "forces", "stress"]
-
-
-def evaluate_autopara_wrappable(
-    atoms,
-    workdir_root=None,
-    dir_prefix="run_VASP_",
-    calculator_command=None,
-    calculator_kwargs=None,
-    output_prefix="VASP_",
-    properties=None,
-    keep_files="default",
-):
-    """Evaluate a configuration with VASP
+_default_keep_files = ["POSCAR", "INCAR", "KPOINTS", "OUTCAR", "vasprun.xml", "vasp.out"]
+_default_properties = ["energy", "forces", "stress"]
+
+_vasp_kwargs_def = {
+    "isif": 2,
+    "isym": 0,
+    "nelm": 300,
+    "ismear": 0,
+    "sigma": 0.05,
+    "ediff": 1.0e-7,
+    "lwave": False,
+    "lcharg": False,
+    "pp": "."
+}
+
+class Vasp(ase.calculators.vasp.vasp.Vasp):
+    """Extension of ASE's Vasp calculator that can be used by wfl.calculators.generic
+
+    Notes
+    -----
+    "directory" argument cannot be present. Use rundir and workdir instead.
```

Comment:
Would it help if I merged `workdir and `rundir`? There's no reason why `rundir` can't be a full path. The reason it's this way is to facilitate things like the gap-rss CLI, where you want to say what directory to put all the per-config run dirs in, but don't necessarily care about their names. But it'd be easy enough to specify, and I agree that rundir vs workdir makes things more confusing.

Reply:
I think "want to say what directory to put all the per-config run dirs in, but don't necessarily care about their names" would be true in most cases (definitely for me). I don't think merging would make it any easier, but naming might clear things up. How about `workdir`, `per_config_rundir_prefix` and either `reuse_rundir_path` without `reuse_rundir` or `per_config_rundir` with `reuse_rundir`?  And if `reuse_rundir_path` or `per_config_rundir` is specified, that's used instead of `workdir` and `per_config_rundir_prefix`.

Reply:
I see what you're getting at, and I think it's generally a good idea, but I want to think a bit more about the exact details.    I don't like having arguments that cause others to be ignored - I think it's too easy to forget which one overrides.

Reply:
What do you think about having `workdir` and a mutually exclusive pair `per_config_rundir_prefix` and `rundir`? If the `*rundir*` are absolute paths, they're used as is, while if they are relative paths, they are relative to `workdir`.  `workdir` can default to `"."`, and we can complain if you pass an explicit `workdir` and an absolute path for `*rundir*`.

Reply:
I initially thought that if `rundir` is to be reused, it's an absolute path rather than with respect to `workdir` (so `workdir` would be ignored). I think not ignoring arguments would be good, so how about just renaming `rundir` to `per_config_rundir` as a reminder about which is which? 

I originally thought that having `rundir` work as a prefix and a path is a bit confusing, but actually with a docstring that should be fine.

Reply:
> What do you think about having workdir and a mutually exclusive pair per_config_rundir_prefix and rundir? If the *rundir* are absolute paths, they're used as is, while if they are relative paths, they are relative to workdir. workdir can default to ".", and we can complain if you pass an explicit workdir and an absolute path for *rundir*.

That would work too, I don't mind which one.

Reply:
So you mean just keeping the current behavior, but renaming `rundir` to `per_config_rundir`, and making sure the dosctring is good? Is it weird that `per_config_rundir` isn't really "per-config" if it's a full name and not a prefix?

Reply:
wait, what happens if we set current `rundir` and parallelise the calculator? All copies are executed in `rundir`? That will crash!

Reply:
Yes, that's true.  We could warn in the dosctring, or we could just drop the non-prefix syntax altogether.  It is potentially useful, but not often, and only when sequential configs are related, which contradicts the basic assumption of the workflow that different configurations are independent.

Reply:
can we and how tricky would it be to just refuse to parallelise (at all or more than one-way) with `generic` if a calculator has `reuse_rundir` and it is set to `True`?

Reply:
Let me look at the code.  My instinct was that it's really ugly to depend on such internal details outside of the calculator, but if we've refactored those arguments into a `wfl` utility, it's not that weird to have functions outside of the DFT calculators know about it, and respond. 

Clearly we need to refuse if it's using `multiprocessing.pool` parallelization.  Remote job parallelization is actually OK as long as the rundirs are relative to each job, i.e. not absolute paths.  Are we willing to make the logic that complex?  Or is that a sign that we should drop support for this altogether?

Reply:
hmm I think maybe it would be best to drop the support for this. We can revisit this if there's a use-case at hand, otherwise launching ASE's VASP in the directory would do the job (even if it would be somewhat inconvenient to go back and re-set up ASE's VASP. But supporting this is turning out to be even less alluring.)

Reply:
OK, I'll remove it, and leave a comment that we might consider re-adding it if it seems useful in the future.

Reply:
dropping this problematic feature.